src/tiled/bucketfilltool.cpp

```diff
@@ -360,6 +391,59 @@ void BucketFillTool::randomFill(TileLayer &tileLayer, const QRegion &region) con
     }
 }
 
+//Helper function which gets a cell from the front or background tilelayer based on x and y
+//All positions are given relative to the frontTL (and assumes backTL.pos = 0,0)
+//If outside fillRegion, then a cell is given from backTL (if it has one)
+//Wasent sure what to name this, so sloppy names for now.
+static const Cell &adjCell(const TileLayer *backTL, const TileLayer *frontTL, const QRegion &fillRegion, int x, int y)
+{
+    if (!fillRegion.contains(QPoint(x + frontTL->x(), y + frontTL->y())) && backTL->contains(x + frontTL->x(), y + frontTL->y())) {
+        return backTL->cellAt(x + frontTL->x(), y + frontTL->y());
+    } else if (frontTL->contains(x, y)) {
+        return frontTL->cellAt(x, y);
+    } else {
+        const Cell &cell = Cell();
+        return cell;
+    }
+}
+
+void BucketFillTool::wangFill(TileLayer &tileLayerToFill,
+                              const TileLayer &backgroundtileLayer,
+                              const QRegion &region) const
+{
+    if (region.isEmpty() || !mWangSet)
+        return;
+
+    for (const QRect &rect : region.translated(-tileLayerToFill.position()).rects()) {
+        for (int _x = rect.left(); _x <= rect.right(); ++_x) {
+            for (int _y = rect.top(); _y <= rect.bottom(); ++_y) {
+                tileLayerToFill.setCell(_x, _y, Cell());
```

Comment:
We should probably add a `TileLayer::clear` method, but let's not do that now because its implementation will need to be different on `master`. So we should wait with that until after the merge.